### DISCUSSION TITLE
Echo2: A Tiny Piece of Code for Learning and Practicing Soroban Rust SDK

### DISCUSSION THREAD

**Discussion by huitemagico** [stellar/sorobounty-spectacular#15] [General]
Hello everyone,

To facilitate the learning process of the Soroban SDK and provide to the beginners, like myself, with an example, I have created a program called "**echo2**". 

### Function description
This program takes a text parameter as input from the caller, retrieves the previous message from **persistent storage**, and then saves the new message. 
So, with each subsequent call, echo2 returns "the echo of your message".
An easy way of explain the function of echo2, please see next paragraph.

You can find the code and documentation at: (https://github.com/huitemagico/echo2)

I post documentation also in  dev.to site [https://dev.to/huitemagico/echo2-a-tiny-piece-of-code-for-learning-and-practicing-soroban-rust-sdk-4h2m](https://dev.to/huitemagico/echo2-a-tiny-piece-of-code-for-learning-and-practicing-soroban-rust-sdk-4h2m)

### The finite state machine of echo2
Echo2 is a program that implements a basic state machine with two states: reset_state and echo_state. In the reset_state, the program does not process the previous message but instead resets it with a special meaning. If Echo2 receives another reset message while in the reset_state, it remains in that state. However, if it receives a message different from "reset", it transitions to the echo_state while saving the previous state and counter.

While in the echo_state, if Echo2 receives a general message that is different from "reset", it remains in that state. The machine continues to stay in the echo_state until it receives a "reset" message, at which point it transitions back to the reset_state.

The explanation can be complex, which is why it is recommended to design the state machine using a "finite state machine diagram" for better understanding. Please see the diagram below.
![echo2 as finite state machine](https://github.com/stellar/sorobounty-spectacular/assets/1184319/509e1244-6ba0-4ccf-8958-3790450862c0)

### My vision related to this program
In my opinion, and considering my own beginner level of expertise, I believe there is a significant opportunity for creating **more basic examples, with documentation**,  and with that, making the code writing process with the SDK much easier.
Specifically, I consider echo2 is one example of a set of components (subprograms, routines) that I intend to build (and discover!) aimed at facilitating the programming of smart contracts.

### About me.
Nowadays I am a very happy freelance programmer.
 I  began with punched cards programming with UR machines, then Assembler IBM 360 in the seventys.
 My experience include PL/1, Algol, Fortran, COBOL, C.  
 After working as programmer in the beginning, I was Business Analyst for a lot (too much!)of years.
Lasts years I had the opportunity for working hard on Java Spring Batch project.
 And now obviously I am converted to Soroban fan :-D , because I think that Smart Contract is a new big revolution.

**Comment by vinamogit** [comment]
You get the echo of the previous caller.
This is funny,  and in some ways it could be turned into a [cadavre exquis](https://en.wikipedia.org/wiki/Exquisite_corpse)

**Comment by esteblock** [comment]
Cool project, congrats!

**Comment by briwylde08** [comment]
Hello! Thank you so much for your submission to Sorobounty Spectacular! We will review it and get back to you within a week. 

Best,
Bri

**Comment by ElliotFriend** [comment]
# Hi, @huitemagico!!

## TL;DR

- Really cool submission!
- I'm curious about something you mentioned about persistent storage in production
- There's a few small changes I'd love to see
- Try and run through your writing to fix spelling/grammar issues, if you've got time.

## My Impression

You've got a really creative, fun use of Soroban here! And, I think you've identified a spot for potential improvement in our documentation: the basic, "small" examples that can help _brand new_ users/devs get up and running. Really cool! Thanks for submitting it!!

## Questions

I noticed something that made me curious. From your README file, in section (4):

> note: persistent storage is not recomended for production environment!

I'm curious what brought you to that conclusion? I couldn't find anything similar in the links you had nearby.

## A few things I'd love to see changed

1. When I first looked at the repository structure, I was a bit confused by the "additional" files I found (like `compi.sh`, `run1.sh`, `myhistorial.txt`, etc.). It's pretty interesting to see the progression of your experimentation, but I'd want to see those files cleaned out of the repo, or maybe just consolidated into a single `something.sh` file that keeps a record where you could copy/paste from.
2. While the Rust code you've written is pretty concise and easy to follow, it could use some cleanup and more comments inside the code. Along those lines, you can use `cargo fmt` to automatically format your source code along the typical formatting conventions used in Rust.
3. The `README.md` file could use just a bit of rearranging. I think it would be really good to have the section "(5) Steps for run the program" earlier in the file. Some of the parts you've got in code blocks don't really need to be that way. The links you share in section (4) are really useful, but there's some strange indentation happening.
4. The article you've written for dev.to is a good collection of links/resources/concept, but I'd love for it to be a bit more "tutorial-like" as far as _how_ you got the project working.

## One non-blocking thing

Some of the spelling/grammar could use some going over. This one is less of a deal (to me), though. I know English isn't everyone's first language, and almost everything I noticed in your submission might just be a result of that. If you can give your stuff another once-over for that kind of stuff, it would help take the submission to the next level 👍🏻

**Comment by huitemagico** [comment]
Hello Elliot!
I sincerely thank you for the feedback. I will work on the points you
indicate. Apologies for the delay in answering, I am right now in an almost
"off grid" location. Now I have a three-day driving trip to get home, so I
please request some time to resolve the pending issues. Some notes in this
regard:
a) Regarding the specific issue of the use of "persistent storage" in a
production environment, I have been reviewing the documentation right now
and I cannot find the precise phrase either. :-/ It may be my
misunderstanding.
b) of course the repository requires an order! I will do it as soon as
possible.
c) I will work with rust fmt and improvements in comments in the source
program
d) regarding the doc in dev.to, I will give it an order and content for a
novice like me, clarifying the points in which I lost a lot of time, due to
lack of adequate knowledge.
e) and sorry for my English, I will check that too.

Greetings.

**Reply by ElliotFriend** [reply]
No problem at all! Hope you're having a safe trip home! I was off for the weekend, too, so no worries 👍🏻 

> a) ... I have been reviewing the documentation right now and I cannot find the precise phrase either. ...

If you come across anything in the docs that strikes your memory as possibly having led you to that, let me know. We can definitely clarify/change anything if we need to.

> d) ... I will give it an order and content for a novice like me ...

That sounds perfect!! Exactly what we'd love to see

> e) and sorry for my English, I will check that too.

No apologies necessary here! Like I mentioned, this is a non-blocking issue. Just a quick run of spell/grammar check should more than take care of it.

Thanks again for your submission, and the hard work you're putting into it! Let me know if you come across any trouble, or need help with anything.

Thanks!!

**Comment by huitemagico** [comment]
Thank you, you too!

**Reply by huitemagico** [reply]
Hello Elliot !
I updated the README.md on GitHub https://github.com/huitemagico/echo2 
and also wrote a "tutorial-style" documentation with all the steps 
to install all the necessary components to run the program. 
The "tutorial" is on GitHub at 
https://github.com/huitemagico/echo2/blob/rama1/echo2tutorial.pdf.

I also updated dev.to at https://dev.to/huitemagico/echo2-a-tiny-piece-of-code-for-learning-and-practicing-soroban-rust-sdk-4h2m

I hope that these updates meet your expectations. Greetings, and thank you for reading.

**Reply by ElliotFriend** [reply]
Hi, there @huitemagico!!

That tutorial you wrote is fantastic!! It's really great having the whole process from beginning to end from a beginner's perspective. The README, tutorial, and dev.to article all look great!

Thanks for the hard work you put in on this project. Very well done!

Thanks!

**Reply by esteblock** [reply]
Wow! I think this is one of the best tutorials we have so far for Soroban begginers! @briwylde08 @Julian-dev28

**Comment by briwylde08** [comment]
Hello @huitemagico! Thank you so much for your submission and discussion. We've reviewed your submission! [🎉](https://emojipedia.org/party-popper/)

Given our guidelines, your submission may qualify for an award of $700 worth of XLM. To receive that award, please fill out [this form](https://stellarform.typeform.com/to/HBjEyCJt) completely and accurately.

Thanks again! Please let us know if you have any questions.

Best,
Bri

**Reply by huitemagico** [reply]
Hello Bri! Thank you very much!   :-D
I am very grateful for the opportunity to work in the community. 
Furthermore, I really appreciate that communication is done in such a human and friendly manner.
I will follow the indicated procedure.
Greetings

**Reply by huitemagico** [reply]
Hello Bri! I have just finished the procedure, including filling out the forms, and I have also sent my wallet address. I hope everything is okay, so I am waiting for news from you. Thanks and regards.

**Reply by briwylde08** [reply]
Hi there! I'm so sorry if you've been waiting for a while. I was at our conference in Madrid last week and then on vacation this week. I am back working tomorrow so will make sure to process everything then. Thank you!!

**Reply by huitemagico** [reply]
No worries, thanks to you!

**Reply by briwylde08** [reply]
Hello @huitemagico! Would you mind checking your email?

**Comment by huitemagico** [comment]
Hello! please give me your discord is easier follow the chat that way, regards
Regards,

**Comment by briwylde08** [comment]
Hi @huitemagico! Thank you for providing the requested information. You’ve passed KYC and we’ve logged your award payment of $700 worth of XLM. This should be sent to your wallet within the next week.

Please let me know if you have any questions!

Best,
Bri

**Reply by huitemagico** [reply]
Hello, thanks so much. Regards,
